source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc

```diff
@@ -123,17 +128,61 @@ bool LocalRateLimiterImpl::requestAllowedHelper(const TokenState& tokens) const
   return true;
 }
 
-bool LocalRateLimiterImpl::requestAllowed(
+StatusOr<std::reference_wrapper<const LocalRateLimiterImpl::LocalDescriptorImpl>>
+LocalRateLimiterImpl::descriptorHelper(
     absl::Span<const RateLimit::LocalDescriptor> request_descriptors) const {
   if (!descriptors_.empty() && !request_descriptors.empty()) {
+    // The override rate limit descriptor is selected by the first full match from the request
+    // descriptors.
     for (const auto& request_descriptor : request_descriptors) {
       auto it = descriptors_.find(request_descriptor);
       if (it != descriptors_.end()) {
-        return requestAllowedHelper(*it->token_state_);
+        return *it;
       }
     }
   }
-  return requestAllowedHelper(tokens_);
+  return absl::Status(absl::StatusCode::kNotFound, "no descriptor found");
+}
+
+bool LocalRateLimiterImpl::requestAllowed(
+    absl::Span<const RateLimit::LocalDescriptor> request_descriptors) const {
+  auto descriptor = descriptorHelper(request_descriptors);
+
+  return descriptor.ok() ? requestAllowedHelper(*descriptor->get().token_state_)
+                         : requestAllowedHelper(tokens_);
+}
+
+uint32_t LocalRateLimiterImpl::maxTokens(
+    absl::Span<const RateLimit::LocalDescriptor> request_descriptors) const {
+  auto descriptor = descriptorHelper(request_descriptors);
+
+  return descriptor.ok() ? descriptor->get().token_bucket_.max_tokens_ : token_bucket_.max_tokens_;
+}
+
+uint32_t LocalRateLimiterImpl::remainingTokens(
+    absl::Span<const RateLimit::LocalDescriptor> request_descriptors) const {
+  auto descriptor = descriptorHelper(request_descriptors);
+
+  return descriptor.ok() ? descriptor->get().token_state_->tokens_.load(std::memory_order_relaxed)
+                         : tokens_.tokens_.load(std::memory_order_relaxed);
+}
+
+int64_t LocalRateLimiterImpl::remainingFillInterval(
+    absl::Span<const RateLimit::LocalDescriptor> request_descriptors) const {
+  using namespace std::literals;
+
+  auto current_time = time_source_.monotonicTime();
+  auto descriptor = descriptorHelper(request_descriptors);
+  if (descriptor.ok()) {
+    ASSERT(std::chrono::duration_cast<std::chrono::milliseconds>(
+               current_time - descriptor->get().token_state_->fill_time_) <=
+           absl::ToChronoMilliseconds(descriptor->get().token_bucket_.fill_interval_));
+    return absl::ToInt64Seconds(
+        descriptor->get().token_bucket_.fill_interval_ -
+        absl::Seconds((current_time - descriptor->get().token_state_->fill_time_) / 1s));
+  }
+  return absl::ToInt64Seconds(token_bucket_.fill_interval_ -
+                              absl::Seconds((current_time - tokens_.fill_time_) / 1s));
 }
```

Comment:
Some comments would be helpful here, it's a bit hard to parse what's going on

Reply:
Ok, I will add comment here.